### ISSUE TITLE
Play command flags

### ISSUE THREAD

**Issue by rafaeldamasceno** [aquelemiguel/parrot#51]
## Rationale

When playing a playlist, you sometimes want to shuffle the playlist before playing it. Sometimes you might even want to play it right after the current track, or even interrupt the current track and skip right away. These can be all achieved via flags and it was a much appreciated feature from [Groovy](https://groovy.zendesk.com/hc/en-us/articles/360031046952-Advanced-Usage-Flags).

## Description

When using the play command, adding a flag modifies the behavior of the play command.

These were the available flags in Groovy (with the exception of **-c**hoose, which seems more appropriate as a search command). They were also available as shorthand flags with the first letter only:

- [x] **-a**ll: This instructs Groovy to queue all returned songs (...) when you queue a YouTube video in a playlist.
- [x] **-s**huffle: (...) randomize the order of songs in the playlist before it's added to the queue (...)
- [x] **-n**ext: (...) the song you queued is placed directly after the playing track in the queue. (...)
- [x] **-j**ump: Instantly jumps to the track you just queued. (...)
- [x] **-r**everse: Queue the playlist in reverse order.

**Comment by rafaeldamasceno** [comment]
I'd like to return to this feature, since it's the one I miss the most and I want something to start out with the code as well. Do you think it makes sense to keep the flags in the query, or should I make these optional booleans in the slash command?

**Comment by afonsojramos** [comment]
Definitely adding these as optional booleans is the way to go moving forward for any additional arguments imo!

**Comment by aquelemiguel** [comment]
@rafaeldamasceno I reckon this is a good chance to use [select menus](https://discord.com/developers/docs/interactions/message-components#select-menus) as they support both single and multiple selections.
Was it possible to use multiple flags on Groovy?

**Comment by rafaeldamasceno** [comment]
> Was it possible to use multiple flags on Groovy?

Yes, it was.

> @rafaeldamasceno I reckon this is a good chance to use [select menus](https://discord.com/developers/docs/interactions/message-components#select-menus) as they support both single and multiple selections.

Yes and no... Do we want to break the current flow of just posting the query and having the bot reply? We'd need to have this intermediate step to select the flags (which I don't even know if is possible to do in the same slash command as the one that receives the query), or create a new command for using the flags. None of these options seem good to me.

**Comment by afonsojramos** [comment]
@rafaeldamasceno have you tested it out? I'm pretty sure it is pretty seamless using discord's built-in optionals and stuff

**Comment by rafaeldamasceno** [comment]
@afonsojramos No, but I'm pretty sure that select menus are not part of the command options, but instead part of the message components. This would mean having to use `play` (or any other command for this purpose), having the client render the select menu, and only then we would have the correct flags. This contrasts versus using play and having all the flags as optional boolean application command options, and when you hit enter, you are done.

**Comment by aquelemiguel** [comment]
@rafaeldamasceno You're absolutely right, I was confusing that with [application command option choices](https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-option-choice-structure).
Unsure of the limitations (this is still quite new) but maybe something like this? 🤷‍♂️

![image](https://user-images.githubusercontent.com/5618493/150802388-b0652b74-49cc-46aa-a413-f4b002321021.png)

**Comment by aquelemiguel** [comment]
@rafaeldamasceno Do you want me to assign you to this topic?

**Comment by rafaeldamasceno** [comment]
> @rafaeldamasceno Do you want me to assign you to this topic?

Yes, please.

**Comment by aquelemiguel** [comment]
Just an FIY, after #135 we **may** use custom [yt-dlp arguments](https://github.com/yt-dlp/yt-dlp#usage-and-options) to help us implement some flags.
Could be handy, but haven't checked in detail.

**Comment by aquelemiguel** [comment]
@rafaeldamasceno As discussed, moving the issue to @StaticRocket.

**Comment by StaticRocket** [comment]
If @rafaeldamasceno is fine with it then sure!

**Comment by rafaeldamasceno** [comment]
@aquelemiguel Regarding the examples you gave for the options I presented in #140, they are both correct.

The only feasible ways I see to implement this are two: 

- The first one is to have it as booleans like `/play query:query all:true shuffle:true` where the all flags are optionals and the ones you want to activate need to be set to `true`, since it's the closest thing to flags (and unfortunately it's not looking like it will [change](https://github.com/discord/discord-api-docs/issues/2329) [anytime](https://github.com/discord/discord-api-docs/issues/2331) [soon](https://github.com/discord/discord-api-docs/discussions/3286). I don't think it is cumbersome to set, but I agree it looks atrocious. The advantage is that there's helper text and it's obvious to the user what is going to happen when setting values to the flags.
- The alternative would be `/play query:query flags:as`, where the `flags` parameter will have letters for each flag you want to activate. Since `flags` would be an optional, there isn't any need to add dashes or anything. For me, this is the way that would make it fastest for power users to interact with the bot and now that I think about it, is about the same way as Rythm worked and advertised these features (as *advanced usage* flags).

**Comment by joao-conde** [comment]
All different play modes implemented, closing.

**Comment by aquelemiguel** [comment]
Keeping it closed for now. If needed, let's open a separate issue regarding multiple flags for tracking. But realistically, we must wait until Discord supports some sort of [variadic arguments](https://github.com/discord/discord-api-docs/discussions/3286) to justify any changes.
